examples/tabular_q.py

```diff
@@ -0,0 +1,169 @@
+# Copyright (c) Facebook, Inc. and its affiliates.
+#
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+
+"""Simple compiler gym tabular q learning example.
+Usage python tabular_q.py
+
+Using selected features from Autophase observation space, given a specific training
+program as gym environment, find the best action sequence using online q learning.
+"""
+
+import random
+from time import time
+
+import humanize
+from absl import app, flags
+
+from compiler_gym.util.flags.benchmark_from_flags import benchmark_from_flags
+from compiler_gym.util.flags.env_from_flags import env_from_flags
+
+flags.DEFINE_list(
+    "actions",
+    [
+        "-simplifycfg",
+        "-reg2mem",
+        "-early-cse-memssa",
+        "-gvn-hoist",
+        "-gvn",
+        "-instsimplify",
+        "-instcombine",
+        "-jump-threading",
+        "-loop-extract",
+        "-loop-reduce",
+        "-loop-versioning",
+        "-newgvn",
+        "-mem2reg",
+        "-sroa",
+        "-structurizecfg",
+    ],
+    "A list of action names to explore from.",
+)
+flags.DEFINE_float("discount", 1.0, "The discount factor.")
+flags.DEFINE_list(
+    "features_indices",
+    [19, 22, 51],
+    "Indices of Alphaphase features that are used to construct a state",
+)
+flags.DEFINE_float("learning_rate", 0.1, "learning rate of the q-learning.")
+flags.DEFINE_integer("episodes", 5000, "number of episodes used to learn.")
+flags.DEFINE_float("epsilon", 0.2, "Epsilon rate of exploration. ")
+flags.DEFINE_integer("episode_length", 5, "The number of steps in each episode.")
+FLAGS = flags.FLAGS
+
+
+def hash_state_action(autophase_feature, action, step):
+    return tuple(
+        (
+            tuple(autophase_feature[FLAGS.features_indices]),
+            step,
+            FLAGS.actions.index(action),
+        )
+    )
```

Comment:
In general I would prefer there to be a bit more documentation for the functions in this file. In particular, this one, since, even if the reader is a compiler/rl expert, this function contains both implementation details (the need to create a hashable tuple), and design decisions (the choice of autophase feature values), etc.

I would suggest something like this (though note it is an incomplete scripbble :-) ):

```suggestion
def make_q_table_key(autophase_feature, action, step):
    """Create a hashable Q-table key.

    For tabular learning we will be constructing a Q-table which maps a
    (state, action) pair to an expected reward. The purpose of this function is
    to convert the (state, action) properties into a hashable tuple that can be
    used as a key for a Q-table dictionary.

    The (state, action) tuple returned by this function consists of ... (TODO:
    explain the choice of autophase features, the purpose of step count, etc).
    """
    return *autophase_feature[FLAGS.features_indices], step, FLAGS.actions.index(action)
```

I would point to [finish_episode()](https://github.com/facebookresearch/CompilerGym/blob/development/examples/actor_critic.py#L240-L307) as a guide for figuring out the level of detail required when documenting non-obvious code